TASK: Clean up stored throwable dumps

kdambekalns opened the pull request:
Whenever a new dump is written, check the existing dumps and remove those that are older than allowed or exceed the maximum count.

By default nothing is cleaned up.

Resolves #3158

**Review instructions**

Should remove old dump files as configured…

**Checklist**

- [x] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [x] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [x] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [x] Reviewer - The first section explains the change briefly for change-logs
- [ ] ~Reviewer - Breaking Changes are marked wit `!!!` and have upgrade-instructions~

kdambekalns committed:
TASK: Clean up stored throwable dumps

Whenever a new dump is written, check the existing dumps and remove
those that are older than allowed or exceed the maximum count.

By default nothing is cleaned up.

kdambekalns commented:
I admit:

- I did not run this code once, yet 🙈
- It might look like a feature, but I want this in 7.3, as it can avoid disks being filled

@mrimann – feel free to test, if you dare…

kdambekalns commented:
Fixed PHP 7.3 compatibility for Flow 7.3… 🙄🤣

mrimann commented:
I tried, but failed. I'm not 100% sure if I tested wrong, or if the code does not yet do what I'd have expected.

My test-setup was:
- fresh flow-base-distribution, 7.3.x branch
- in a DDEV environment with PHP 7.4.x
- composer install
- empty DB + applied all migrations
- copied over the 2 changed files from this pull request

Then I tried to provoke Flow into writing exceptions by calling `/foobar` via the browser -> shows an error, but does not write out an Exception / Dump file.

Next I "broke" the DB by removing an entry from the migrations table - re-applying the same migration now throws an error and writes a Dump file.

But no exceptions are removed, even if I configure the `maximumThrowableDumpCount: 2`.

(I verified that editing those settings works by temporarily setting the `storagePath` option to something else -> this immediately worked. I also tried to circumvent any problems with the configuration injection, by adding the line `$this->maximumThrowableDumpCount = 2;` right before the `if ()` that checks if that setting is >0 - but this also does not remove any of the exception files).

@kdambekalns sorry - but hope this helps a bit.

ahaeslich commented on Neos.Flow/Classes/Log/ThrowableStorage/FileStorage.php:
```diff
@@ -233,4 +255,35 @@ protected function renderRequestInfo()
+        if ($this->maximumThrowableDumpCount > 0) {
+            // this returns alphabetically ordered, so newest first, as we have a date/time in the name
+            $existingDumps = glob($this->storagePath);
+            if (count($existingDumps) > $this->maximumThrowableDumpCount) {
```
`count($existingDumps)` will always return 1 as `$this->storagePath` contains the folder (e.g. `/myFlowPath/Data/Logs/Exceptions`) instead of a glob pattern (e.g. `/myFlowPath/Data/Logs/Exceptions/*`).

kdambekalns replied:
Oh my… I actually remember writing this in my head:
```php
$existingDumps = glob($this->storagePath . '/*');
```
and then stopping because I wondered about `$this->storagePath` having a trainling slash or not and how `glob` orders things. That got me distracted enough, and I forgot…

Thanks for seeing this! ❤️

ahaeslich replied:
> I actually remember writing this in my head

A classic. I definitly can relate to that 😄

ahaeslich reviewed (changes requested):
I tested it in a 8.3 installation and using `maximumThrowableDumpAge` worked as expected.

ahaeslich commented on Neos.Flow/Classes/Log/ThrowableStorage/FileStorage.php:
```diff
@@ -233,4 +255,35 @@ protected function renderRequestInfo()
+            // this returns alphabetically ordered, so newest first, as we have a date/time in the name
+            $existingDumps = glob($this->storagePath);
+            if (count($existingDumps) > $this->maximumThrowableDumpCount) {
+                $dumpsToRemove = array_slice($existingDumps, $this->maximumThrowableDumpCount);
```
On second view: did you want to remove files in this case regardless of their creation date?

kdambekalns replied:
Yes, I don't care at that point. Because: The dumps files

- are ordered by name, see , where it says
   > GLOB_NOSORT -  […]. When this flag is not used, the pathnames are sorted alphabetically 
- and the name starts with a timestamp generated through `date('YmdHis', $time)`
- so the older files come first

Which makes the code almost correct… 😇 The maximum should be left, not removed, so it should be

```php
$dumpsToRemove = array_slice($existingDumps, -$this->maximumThrowableDumpCount);
```

to slice from the end of the array, see https://3v4l.org/FUGWC

kdambekalns replied:
And now of course those are the ones that should **not** be deleted. 🤪

kdambekalns replied:
Like this: https://3v4l.org/46LTA

kdambekalns commented:
@mrimann wrote:
> But no exceptions are removed, even if I configure the `maximumThrowableDumpCount: 2`.

As @ahaeslich pointed out, exactly that limit was implemented wrong… Now it should work!

ahaeslich reviewed (approved):
> It might look like a feature, but I want this in 7.3, as it can avoid disks being filled

One can easily argue that it is a bug that disks are filled with exceptions without a cleanup mechanism😉. No objection on my part on targeting 7.3.

Btw: `maximumThrowableDumpCount` works now 👍

kitsunet reviewed (changes requested):
I am against putting this into a released version, it changes the constructor of the FileStorage, api or not, this might very well be used in userland code and breaking it in a bugfix release is a nogo.

mrimann commented:
Thanks @kdambekalns for the second version of the change. I could now successfully test it's behaviour (with the same test-setup as outlined in the previous comment). The only functional thing I noticed is that setting `maximumThrowableDumpCount` to e.g. "2", results in 3 Dump files remaining on the disk (the two allowed ones + the new one that is written after cleaning up). But I think that's OK and still improves the situation and avoids filling up the disk with dump files.

Regarding the branch to target, I'm with @kitsunet and against adding this to older releases as it adds new config options - and the new feature of cleaning up is off by default (so just updating to the latest patchlevel release "won't solve a problem" (without manually changing some settings).

Instead I'd prefer to see a solution to the second issue https://github.com/neos/flow-development-collection/issues/3159 that I've initially brought up, backported to Neos 7.x (once there's a solution) instead of this here.

kdambekalns commented:
@kitsunet It changes what the constructor does internally, in a very careful way. What kind of breaking behavior do you have in mind? 🤔

@mrimann So far I did not pick "sane defaults" precisely to not change the behavior, I'd love to, though. Just not sure what would be good and somehow match the log file rotation defaults. 🤷

About the "off-by-one, true… But with anything in real-life that probably doesn't matter. 😇

mrimann commented:
@kdambekalns who knows what "sane" is :-) IMO good defaults ("good" as in "acceptable by large parts of the community") could be to
- remove dumps older than 30 days (if they have not been looked after within 30 days, no one ever will)
- remove older dumps in case there are >10'000 dumps on the disk (a high limit, but not that high that the disk will fill up from the dump files)

kdambekalns committed:
TASK: Update default settings

This updates the default settings in YAML to 30 days of dump retention and a maximum of 10.000 files.

The class properties keep their `0` default, so that in case the class has been extended no change is enforced.

ahaeslich commented:
@kdambekalns & @kitsunet which version do we want to target now?

kitsunet committed:
Update Settings.Log.yaml

Revert settings to make this per default a no-op in released versions.

kitsunet reviewed (approved):
Lets merge it like this.

bwaidelich commented on Neos.Flow/Configuration/Settings.Log.yaml:
```diff
@@ -58,3 +58,7 @@ Neos:
         optionsByImplementation:
           'Neos\Flow\Log\ThrowableStorage\FileStorage':
             storagePath: '%FLOW_PATH_DATA%Logs/Exceptions'
+            # The maximum age of throwable dump in seconds, 0 to disable cleaning based on age, default 30 days
```
> default 30 days

is that true?

ahaeslich replied:
We discussed at the sprint to disable this feature by default to not change the behaviour in a path release. But I guess we missed changing the comments 🙈

bwaidelich commented on Neos.Flow/Configuration/Settings.Log.yaml:
```diff
@@ -58,3 +58,7 @@ Neos:
             storagePath: '%FLOW_PATH_DATA%Logs/Exceptions'
+            # The maximum age of throwable dump in seconds, 0 to disable cleaning based on age, default 30 days
+            maximumThrowableDumpAge: 0
+            # The maximum number of throwable dumps to store, 0 to disable cleaning based on count, default 10.000
```
> default 10.000

And that? I don't see any code for this default behavior

ahaeslich replied:
See here. It was the original default and will be introduced in Neos 9: https://github.com/neos/flow-development-collection/pull/3213

bwaidelich replied:
Thanks for the update (and for the record: comments for 7.3-8.x will be fixed with #3269)
